### PR TITLE
ALC-145 Sign approval & swap transactions on client

### PR DESCRIPTION
**Pull request by swgz9**

### CODE REVIEW

File: src/hooks/useApproveCallback.ts [hunk 1]
```diff
@@ -24,8 +27,8 @@ export enum ApprovalState {
 export function useApproveCallback(
   amountToApprove?: CurrencyAmount,
   spender?: string
-): [ApprovalState, () => Promise<void>] {
-  const { account } = useActiveWeb3React()
+): [ApprovalState, () => Promise<string | undefined>] {
```
Comment (CodeForcer) [inline src/hooks/useApproveCallback.ts:30]: Should we be returning undefined and continuing code flow, or just throwing error if we have an edge case?

File: src/hooks/useApproveCallback.ts [hunk 2]
```diff
@@ -46,55 +49,104 @@ export function useApproveCallback(
...
+    }
 
-    return tokenContract
-      .approve(spender, useExact ? amountToApprove.raw.toString() : MaxUint256, {
-        gasLimit: calculateGasMargin(estimatedGas)
+    //use populate instead of broadcasting
+    return tokenContract.populateTransaction
+      .approve(spender, amountToApprove.raw.toString(), {
+        nonce: tokenContract.signer.getTransactionCount(),
+        gasLimit: calculateGasMargin(estimatedGas) //needed?
```
Comment (CodeForcer) [inline src/hooks/useApproveCallback.ts:120]: We can probably just hardcode this, approve isn't going to have much variation

File: src/hooks/useApproveCallback.ts [hunk 3]
```diff
@@ -46,55 +49,104 @@ export function useApproveCallback(
...
-      .then((response: TransactionResponse) => {
-        addTransaction(response, {
-          summary: 'Approve ' + amountToApprove.currency.symbol,
-          approval: { tokenAddress: token.address, spender: spender }
-        })
+      .then((response: PopulatedTransaction) => {
+        delete response.from
+        response.chainId = chainId
+        const serialized = ethers.utils.serializeTransaction(response)
+        const hash = keccak256(serialized)
```
Comment (CodeForcer) [inline src/hooks/useApproveCallback.ts:126]: I assume you've tested this already, but just so you know hashing functions in JS in general can be problematic to types. Often the function will work with HexString, Buffer, and ArrayBuffers, but give different results for each type

File: src/hooks/useApproveCallback.ts [hunk 4]
```diff
@@ -46,55 +49,104 @@ export function useApproveCallback(
...
+            //   approval: { tokenAddress: token.address, spender: spender }
+            // })
+          })
+          .finally(() => {
+            if (web3Provider) {
+              web3Provider.provider.isMetaMask = isMetamask
+            }
+          })
       })
       .catch((error: Error) => {
```
Comment (CodeForcer) [inline src/hooks/useApproveCallback.ts:145]: We should differentiate between the user rejecting the signing operation, and other types of errors

**Review by CodeForcer — APPROVED** [review]
Looks good to me in terms of code quality. Once we put it on staging we will still need to go through handling the various error cases in the UX.

Not sure how this will play with other wallets, but looks good for MetaMask

File: src/hooks/useApproveCallback.ts (hunk 1 above)
Comment (Dire-0x) [reply src/hooks/useApproveCallback.ts:30]: In the cases where an approval is not needed this callback returns undefined

**Pull request merged**